tooling/image-sync/sync.go

```diff
@@ -0,0 +1,197 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/containers/image/copy"
+	"github.com/containers/image/docker"
+	"github.com/containers/image/signature"
+	"github.com/containers/image/types"
+	"github.com/spf13/viper"
+)
+
+// SyncConfig is the configuration for the image sync
+type SyncConfig struct {
+	Repositories   []string
+	NumberOfTags   int
+	QuaySecretFile string
+	AcrRegistry    string
+	TenantId       string
+	RequestTimeout int
+	AddLatest      bool
+}
+
+// QuaySecret is the secret for quay.io
+type QuaySecret struct {
+	BearerToken  string
+	PullUsername string
+	PullPassword string
+}
+
+// NewSyncConfig creates a new SyncConfig from the configuration file
+func NewSyncConfig() *SyncConfig {
+	var sc *SyncConfig
+	v := viper.GetViper()
+	v.SetDefault("numberoftags", 10)
+	v.SetDefault("requesttimeout", 10)
+	v.SetDefault("addlatest", false)
+
+	if err := v.Unmarshal(&sc); err != nil {
+		Log().Fatalw("Error while unmarshalling configuration %s", err.Error())
+	}
+	Log().Debugw("Using configuration", "config", sc)
+	return sc
+}
+
+// Copy copies an image from one registry to another
+func Copy(ctx context.Context, dstreference, srcreference string, dstauth, srcauth *types.DockerAuthConfig) error {
+	policyctx, err := signature.NewPolicyContext(&signature.Policy{
+		Default: signature.PolicyRequirements{
+			signature.NewPRInsecureAcceptAnything(),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	src, err := docker.ParseReference("//" + srcreference)
+	if err != nil {
+		return err
+	}
+
+	dst, err := docker.ParseReference("//" + dstreference)
+	if err != nil {
+		return err
+	}
+
+	_, err = copy.Image(ctx, policyctx, dst, src, &copy.Options{
+		SourceCtx: &types.SystemContext{
+			DockerAuthConfig: srcauth,
+		},
+		DestinationCtx: &types.SystemContext{
+			DockerAuthConfig: dstauth,
+		},
+	})
+
+	return err
+}
+
+func readQuaySecret(filename string) (*QuaySecret, error) {
+	secretBytes, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	var secret QuaySecret
+	err = json.Unmarshal(secretBytes, &secret)
+	if err != nil {
+		return nil, err
+	}
+
+	return &secret, nil
+}
+
+func filterTagsToSync(src, target []string) []string {
+	var tagsToSync []string
+
+	for _, srcTag := range src {
+		found := false
+		for _, targetTag := range target {
```

Comment:
can we get rid of the nested loop
e.g. wrap target into a `map[string]whatever` and use it for existance lookups?

Reply:
Great idea, will do it.